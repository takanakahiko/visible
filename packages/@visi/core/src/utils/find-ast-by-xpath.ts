import { Element, Node } from 'domhandler';
import { find } from 'domutils';
import XPathAnalyzer, {
  ABSOLUTE_LOCATION_PATH,
  EQUALITY,
  FUNCTION_CALL,
  NODE_NAME_TEST,
  NODE_TYPE_TEST,
  NUMBER,
  POSITION,
  PredicateNode,
  StepNode,
} from 'xpath-analyzer';

const isSameKind = (a: Node, b: Node) => {
  if (a instanceof Element && b instanceof Element) {
    return a.tagName === b.tagName;
  }

  if (!(a instanceof Element) && !(b instanceof Element)) {
    return true;
  }

  return false;
};

const filterSameKind = (topic: Node, nodes: Node[]) =>
  nodes.filter((node) => isSameKind(topic, node));

const testPredicates = (predicates: PredicateNode[], node: Node) =>
  predicates.every((predicate) => {
    if (
      predicate.type === EQUALITY &&
      predicate.lhs.type === FUNCTION_CALL &&
      predicate.lhs.name === POSITION &&
      predicate.rhs.type === NUMBER &&
      node.parent
    ) {
      const selfIndex = filterSameKind(node, node.parent.childNodes).findIndex(
        (child) => child === node,
      );

      return selfIndex === predicate.rhs.number - 1;
    }

    // eslint-disable-next-line
    console.warn(`Predicate ${predicate.type} is not implemented yet`);
    return false;
  });

const testNodeType = (
  node: Node,
  name: string,
  predicates: PredicateNode[] = [],
) => {
  // node.nodeType is number for some reason...
  if (
    name === 'text' &&
    !(node instanceof Element) &&
    testPredicates(predicates, node)
  ) {
    return true;
  }

  return false;
};

const testNodeName = (
  node: Node,
  name: string,
  predicates: PredicateNode[] = [],
) => {
  if (
    node instanceof Element &&
    node.tagName.toLowerCase() === name &&
    testPredicates(predicates, node)
  ) {
    return true;
  }

  return false;
};

const testNode = ([head, ...tail]: StepNode[]) => (
  node: Node | null,
): boolean => {
  if (head == null || node == null) {
    return true;
  }

  if (
    head.test.type === NODE_NAME_TEST &&
    testNodeName(node, head.test.name, head.predicates)
  ) {
    return testNode(tail)(node.parent);
  }

  if (
    head.test.type === NODE_TYPE_TEST &&
    testNodeType(node, head.test.name, head.predicates)
  ) {
    return testNode(tail)(node.parent);
  }

  return false;
};

export const findASTByXPath = (
  nodes: Node | Node[],
  xpathStr: string,
): Node | undefined => {
  const root = Array.isArray(nodes) ? nodes : [nodes];
  const xpath = new XPathAnalyzer(xpathStr).parse();

  if (xpath.type !== ABSOLUTE_LOCATION_PATH) {
    throw new Error(`Expected absolute xpath, but got ${xpathStr}`);
  }

  const steps = xpath.steps.reverse();
  return find(testNode(steps), root, true, Number.POSITIVE_INFINITY)?.[0];
};
