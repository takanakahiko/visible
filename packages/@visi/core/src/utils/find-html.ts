import { Node } from 'domhandler';
import { findOne } from 'domutils';

export const findHTML = (nodes: Node[]): Node | null => {
  return findOne(
    (element) => {
      return element.tagName === 'html';
    },
    nodes,
    true,
  );
};
