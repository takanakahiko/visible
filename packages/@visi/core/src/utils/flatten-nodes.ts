import { Node } from 'domhandler';
import { getChildren, hasChildren } from 'domutils';

const flattenNode = (node: Node): Node[] => {
  if (!hasChildren(node)) return [node];
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return [node, ...getChildren(node)!.flatMap((node) => flattenNode(node))];
};

export const flattenNodes = (nodes: Node[]): Node[] => {
  return nodes.flatMap((node) => flattenNode(node));
};
