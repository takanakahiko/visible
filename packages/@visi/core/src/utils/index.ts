export * from './find-ast-by-xpath';
export * from './flatten-nodes';
export * from './find-html';
