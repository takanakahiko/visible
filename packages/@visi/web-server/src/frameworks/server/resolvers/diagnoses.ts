import { API } from '../../../interfaces/presenters';
import {
  MutationResolvers,
  QueryResolvers,
  // Report,
  SubscriptionResolvers,
} from '../generated/graphql';

export const rootDiagnosis: QueryResolvers['diagnosis'] = (
  _,
  { id },
  { diagnosisLoader },
) => {
  return diagnosisLoader.load(id);
};

export const createDiagnosis: MutationResolvers['createDiagnosis'] = (
  _,
  { url },
  { diagnosisController },
) => {
  return diagnosisController.create(url);
};

export const deleteDiagnosis: MutationResolvers['deleteDiagnosis'] = (
  _,
  { id },
  { diagnosisController },
) => {
  return diagnosisController.delete(id);
};

export const diagnosisSubscription: SubscriptionResolvers['diagnosis'] = {
  subscribe(_, { id }, { diagnosisController }) {
    return diagnosisController.subscribe(id);
  },
  resolve(diagnosis: API.Diagnosis) {
    return diagnosis;
  },
};
