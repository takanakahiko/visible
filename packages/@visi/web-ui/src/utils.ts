// eslint-disable-next-line @typescript-eslint/no-explicit-any
const pluck = <T extends Record<string, any>>(key: keyof T) => (object: T) =>
  object[key];

const createPluck = <T extends string>(props: T[]) => {
  return props.reduce((plucks, prop) => {
    plucks[prop] = pluck(prop);
    return plucks;
  }, {} as { [key in T]: ReturnType<typeof pluck> });
};

const createShouldForwardProp = (props: unknown[]) => (prop: unknown) =>
  !props.includes(prop);

export interface Hints<T extends string> {
  readonly select: { [key in T]: ReturnType<typeof pluck> };
  shouldForwardProp(prop: unknown): boolean;
}

export const createHelpers = <T extends string>(props: T[]): Hints<T> => ({
  shouldForwardProp: createShouldForwardProp(props),
  select: createPluck(props),
});
