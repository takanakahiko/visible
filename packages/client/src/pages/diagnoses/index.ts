export { Diagnoses as default } from './diagnoses';
