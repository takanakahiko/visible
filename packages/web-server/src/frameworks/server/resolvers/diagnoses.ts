import {
  MutationResolvers,
  QueryResolvers,
  SubscriptionResolvers,
} from '../generated/graphql';

export const rootDiagnosis: QueryResolvers['diagnosis'] = async (
  _,
  { id },
  { diagnosisController },
) => {
  const diagnosis = await diagnosisController.load(id);
  return diagnosis;
};

export const createDiagnosis: MutationResolvers['createDiagnosis'] = async (
  _,
  { url },
  { diagnosisController },
) => {
  const diagnosis = await diagnosisController.create(url);
  return diagnosis;
};

export const deleteDiagnosis: MutationResolvers['deleteDiagnosis'] = async (
  _,
  { id },
  { diagnosisController },
) => {
  return diagnosisController.delete(id);
};

export const diagnosisSubscription: SubscriptionResolvers['diagnosis'] = {
  subscribe(_, { id }, { diagnosisController }) {
    return diagnosisController.subscribe(id);
  },
};
