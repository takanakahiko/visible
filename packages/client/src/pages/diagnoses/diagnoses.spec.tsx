import React from 'react';
import { Switch, Route } from 'react-router';
import { MockedProvider, MockedResponse, wait } from '@apollo/react-testing';
import { act } from '@testing-library/react';
import { render, makeRouteWrapper } from '../../../tests/utils';
import {
  FetchDiagnosisSmallDocument,
  Diagnosis,
  ReportType,
} from '../../generated/graphql';
import { Diagnoses } from './diagnoses';

const mock: MockedResponse[] = [
  {
    request: {
      query: FetchDiagnosisSmallDocument,
      variables: {
        id: '123',
      },
    },
    result: {
      data: {
        diagnosis: {
          __typename: 'Diagnosis',
          id: '123',
          score: {
            __typename: 'Score',
            error: 1,
            warn: 1,
            ok: 0,
          },
          screenshot: '',
          reports: [
            {
              __typename: 'Report',
              id: '456',
              type: ReportType.Ok,
              name: 'img-alt',
              message: 'error',
              xpath: '/html/body',
              html: '<body></body>',
              css: '',
            },
          ],
        } as Diagnosis,
      },
    },
  },
];

const errorMock: MockedResponse[] = [
  {
    request: {
      query: FetchDiagnosisSmallDocument,
      variables: {
        id: '123',
      },
    },
    error: new Error(),
  },
];

describe('Diagnoses', () => {
  it('fetches data', async () => {
    const { container } = render(
      <MockedProvider mocks={mock} addTypename={true}>
        <Switch>
          <Route path="/diagnoses/:id" component={Diagnoses} />
        </Switch>
      </MockedProvider>,
      { wrapper: makeRouteWrapper('/diagnoses/123') },
    );

    await wait(0);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('throws an error', () => {
    const { container } = render(
      <MockedProvider mocks={errorMock} addTypename={true}>
        <Switch>
          <Route path="/diagnoses/:id" component={Diagnoses} />
        </Switch>
      </MockedProvider>,
      { wrapper: makeRouteWrapper('/diagnoses/123') },
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
