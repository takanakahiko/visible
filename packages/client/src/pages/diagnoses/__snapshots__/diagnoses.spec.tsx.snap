// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagnoses fetches data 1`] = `
.c0 {
  box-sizing: border-box;
  width: 1080px;
  margin: 24px auto;
  padding: 12px 18px;
  border-radius: 12px;
  background-color: white;
  box-shadow: 0 0 24px rgba(0,0,0,0.05);
}

<main
  class="c0"
>
  <h1>
    Diagnostics Result: 
    123
  </h1>
</main>
`;

exports[`Diagnoses throws an error 1`] = `
<p>
  Error occurred...
</p>
`;
